src/editor/codemirror/dnd.ts

```diff
@@ -0,0 +1,126 @@
+import { ChangeSet, Transaction } from "@codemirror/state";
+import { EditorView } from "@codemirror/view";
+import { pythonSnippetMediaType } from "../../common/mediaTypes";
+import { calculateChanges } from "./edits";
+
+/**
+ * Information stashed last time we handled dragover.
+ * Cleared on drop of dragleave.
+ */
+interface LastDragPos {
+  /**
+   * The last drag position.
+   */
+  line: number;
+  /**
+   * The inverse set of changes to the changes made for preview.
+   */
+  previewUndo: ChangeSet;
+}
+
+let lastDragPos: LastDragPos | undefined;
```

Comment:
Pretty sure at some point I'd moved this inside the dndSupport so it was per-editor. That seems more defensive to me. I think if there were multiple editors the events would happen in the right order for it not to be an issue, but an easy improvement anyway.